src/ledger/LedgerManagerImpl.cpp

```diff
@@ -911,11 +954,64 @@ void
 LedgerManagerImpl::ApplyState::populateInMemorySorobanState(
     SearchableSnapshotConstPtr snap, uint32_t ledgerVersion)
 {
-    threadInvariant();
+    assertInitPhase();
     mInMemorySorobanState.initializeStateFromSnapshot(
         snap, mSorobanNetworkConfig.get(), ledgerVersion);
 }
 
+void
+LedgerManagerImpl::ApplyState::assertCommitLedgerPhase() const
+{
+    threadInvariant();
+    releaseAssert(mPhase == Phase::COMMIT_LEDGER);
+}
+
+void
+LedgerManagerImpl::ApplyState::markStartOfTxExecution()
+{
+    threadInvariant();
+    releaseAssert(mPhase == Phase::IDLE);
+    mPhase = Phase::TX_EXECUTION;
+}
+
+void
+LedgerManagerImpl::ApplyState::markStartOfCommitLedger()
+{
+    threadInvariant();
+    releaseAssert(mPhase == Phase::TX_EXECUTION);
+    mPhase = Phase::COMMIT_LEDGER;
+}
+
+void
+LedgerManagerImpl::ApplyState::markEndOfCommitLedger()
+{
+    assertCommitLedgerPhase();
+    mPhase = Phase::IDLE;
+}
+
+void
+LedgerManagerImpl::ApplyState::markEndOfInit()
+{
+    threadInvariant();
+    releaseAssert(mPhase == Phase::INIT);
+    mPhase = Phase::IDLE;
+}
+
+void
+LedgerManagerImpl::ApplyState::resetToInitPhase()
+{
+    threadInvariant();
+    releaseAssert(mPhase == Phase::IDLE);
```

Comment:
This looks suspicious, as we shouldn't ever go from idle back to init? Init state should only ever occur on startup when we're setting up. I see that this is used in setLastClosedLedger function, any reason not to be in INIT state there?

Reply:
I added a comment to clarify why we need to call this, but it happens in `setLastClosedLedger`. We call it in startup, but if a node loses sync and needs to catchup again as it's running, we need to re-enter the init phase at the start of the new catchup run.

Reply:
In plain English, what exactly is the INIT state? I've always thought of it as a "booting state," the initial state of the FSM where core loads all its components. Once core moves on, it should never return to INIT. The comment in LedgerManager.h also suggests this. So, something about the abstraction doesn't quite check out. `setLastClosedLedger` is called when we're done applying buckets on startup, which occurs once when starting from scratch. Could we stay in the INIT state while buckets are being applied, then go to IDLE in `setLastClosedLedger`?

Reply:
Ahh, I think I misunderstood some of the catch-up path. When we lose sync, we do go into catch-up, but we just buffer ledgers and don't redownload buckets.

I think the issue is that we "double init" in some history tests. This shouldn't actually happen in prod though, I'll see if I can get the tests to work.